File: bionic-js-java/src/test/java/acceptance/toy1/ToyClass1Test.java
package acceptance.toy1;

import acceptance.toy1.bjs.BjsTestProject;
import acceptance.toy1.bjs.ToyClass1;
import bionic.js.BjsAnyObject;
import bionic.js.Lambda;
import jjbridge.api.value.JSType;
import jjbridge.engine.v8.V8Engine;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.Date;

import static org.junit.jupiter.api.Assertions.*;

public class ToyClass1Test
{
    private Date testDate;
    private Double testDateTimestamp;
    private ToyClass1 bjsObj;

    @BeforeAll
    public static void beforeClass()
    {
        BjsTestProject.setJsEngine(new V8Engine());
    }

    @BeforeEach
    public void before()
    {
        testDateTimestamp = 472867200000D;
        testDate = Date.from(Instant.ofEpochMilli(testDateTimestamp.longValue()));
        bjsObj = new ToyClass1();
        bjsObj.log("any object");
        ToyClass1.evalAndSetValue_static("undefined");
    }

    @Test
    public void testNativeAutoProp()
    {
        assertNull(ToyClass1.nativeAutoProp_static());

        ToyClass1.nativeAutoProp_static("1984!");
        assertEquals("1984!", ToyClass1.nativeAutoProp_static());

        ToyClass1.nativeAutoProp_static(null);
        assertNull(ToyClass1.nativeAutoProp_static());
    }

    @Test
    public void testAnyAutoProp() {
        assertEquals(ToyClass1.anyAutoProp_static().jsObj.getActualType(), JSType.Undefined);

        ToyClass1.anyAutoProp_static(new BjsAnyObject(bjsObj));
        assertEquals("any object", ToyClass1.anyAutoProp_static().getObject(ToyClass1.bjsFactory, ToyClass1.class).log());
    }

    @Test
    public void testBjsObjAutoProp() {
        assertNull(ToyClass1.bjsObjAutoProp_static());

        ToyClass1.bjsObjAutoProp_static(bjsObj);
        assertEquals("any object", ToyClass1.bjsObjAutoProp_static().log());
    }

    @Test
    public void testLambdaAutoProp() {
        assertNull(ToyClass1.lambdaAutoProp_static());

        ToyClass1.lambdaAutoProp_static(string -> string == null ? "null!" : string + " 1984!");
        assertEquals("null!", ToyClass1.lambdaAutoProp_static().apply(null));
        assertEquals("W 1984!", ToyClass1.lambdaAutoProp_static().apply("W"));

        ToyClass1.lambdaAutoProp_static(null);
        assertNull(ToyClass1.lambdaAutoProp_static());

        ToyClass1.lambdaAutoProp_static(string -> null);
        assertNull(ToyClass1.lambdaAutoProp_static().apply("x"));
    }

    @Test
    public void testLambdaAutoProp_putNullToJs() {
        String nullTest = "(this.lambdaAutoProp() === null)";

        ToyClass1.lambdaAutoProp_static(string -> "hey");
        ToyClass1.evalAndSetValue_static(nullTest);
        assertFalse(ToyClass1.boolFunc_static());

        ToyClass1.lambdaAutoProp_static(string -> null);
        ToyClass1.evalAndSetValue_static(nullTest);
        assertTrue(ToyClass1.boolFunc_static());
    }

    @Test
    public void testNativeArrayAutoProp() {
        assertNull(ToyClass1.nativeArrayAutoProp_static());

        ToyClass1.nativeArrayAutoProp_static(new String[][][]{{{"a", "b"}}});
        assertArrayEquals(new String[][][]{{{"a", "b"}}}, ToyClass1.nativeArrayAutoProp_static());

        ToyClass1.nativeArrayAutoProp_static(new String[][][]{{{}, {}}});
        assertArrayEquals(new String[][][]{{{}, {}}}, ToyClass1.nativeArrayAutoProp_static());

        ToyClass1.nativeArrayAutoProp_static(new String[][][]{{}, {}});
        assertArrayEquals(new String[][][]{{}, {}}, ToyClass1.nativeArrayAutoProp_static());

        ToyClass1.nativeArrayAutoProp_static(new String[][][]{});
        assertArrayEquals(new String[][][]{}, ToyClass1.nativeArrayAutoProp_static());

        ToyClass1.nativeArrayAutoProp_static(new String[][][]{null, {null}, {{null}}, {{"test", null}, null}});
        assertArrayEquals(new String[][][]{null, {null}, {{null}}, {{"test", null}, null}}, ToyClass1.nativeArrayAutoProp_static());
    }

    @Test
    public void testBjsObjArrayAutoProp() {
        assertNull(ToyClass1.bjsObjArrayAutoProp_static());

        ToyClass1.bjsObjArrayAutoProp_static(new ToyClass1[][][]{{{bjsObj}}});
        assertArrayEquals(new ToyClass1[][][]{{{bjsObj}}}, ToyClass1.bjsObjArrayAutoProp_static());

        ToyClass1.bjsObjArrayAutoProp_static(new ToyClass1[][][]{{{}, {}}});
        assertArrayEquals(new ToyClass1[][][]{{{}, {}}}, ToyClass1.bjsObjArrayAutoProp_static());

        ToyClass1.bjsObjArrayAutoProp_static(new ToyClass1[][][]{{}, {}});
        assertArrayEquals(new ToyClass1[][][]{{}, {}}, ToyClass1.bjsObjArrayAutoProp_static());

        ToyClass1.bjsObjArrayAutoProp_static(new ToyClass1[][][]{});
        assertArrayEquals(new ToyClass1[][][]{}, ToyClass1.bjsObjArrayAutoProp_static());

        ToyClass1.bjsObjArrayAutoProp_static(new ToyClass1[][][]{null, {null}, {{null}}, {{bjsObj, null}, null}});
        assertArrayEquals(new ToyClass1[][][]{null, {null}, {{null}}, {{bjsObj, null}, null}}, ToyClass1.bjsObjArrayAutoProp_static());
    }

    @Test
    public void testAnyArrayAutoProp() {
        assertNull(ToyClass1.anyArrayAutoProp_static());

        BjsAnyObject anyObj = new BjsAnyObject(bjsObj);

        ToyClass1.anyArrayAutoProp_static(new BjsAnyObject[][][]{{{anyObj}}});
        assertEquals("any object", ToyClass1.anyArrayAutoProp_static()[0][0][0].getObject(ToyClass1.bjsFactory, ToyClass1.class).log());

        ToyClass1.anyArrayAutoProp_static(new BjsAnyObject[][][]{{{}}});
        assertEquals(0, ToyClass1.anyArrayAutoProp_static()[0][0].length);
    }

    @Test
    public void testLambdaArrayAutoProp() {
        assertNull(ToyClass1.lambdaArrayAutoProp_static());

        Lambda.F1<String, String> lambda1a = s -> (s != null ? s : "") + "-1a";
        Lambda.F1<String, String> lambda1b = s -> (s != null ? s : "") + "-1b";
        Lambda.F1<String, String> lambda2a = s -> (s != null ? s : "") + "-2a";

        ToyClass1.lambdaArrayAutoProp_static(new Lambda.F1[][][]{{{lambda1a, lambda1b}, {lambda2a, null}}});

        assertEquals("test-1a", ToyClass1.lambdaArrayAutoProp_static()[0][0][0].apply("test"));
        assertEquals("test-1b", ToyClass1.lambdaArrayAutoProp_static()[0][0][1].apply("test"));
        assertEquals("test-2a", ToyClass1.lambdaArrayAutoProp_static()[0][1][0].apply("test"));
        assertNull(ToyClass1.lambdaArrayAutoProp_static()[0][1][1]);

        ToyClass1.lambdaArrayAutoProp_static(new Lambda.F1[][][]{{{}}});
        assertEquals(0, ToyClass1.lambdaArrayAutoProp_static()[0][0].length);
    }

    @Test
    public void testProp() {
        assertEquals("1984!", ToyClass1.prop_static());

        ToyClass1.prop_static("test value");
        assertEquals("test value", ToyClass1.prop_static());

        ToyClass1.prop_static(null);
        assertEquals("1984!", ToyClass1.prop_static());
    }

    @Test
    public void testVoidFunc() {
        ToyClass1.voidFunc_static();
        assertEquals("called voidFunc", ToyClass1.log_static());
    }

    @Test
    public void testParamsFunc() {
        ToyClass1.paramsFunc_static(true, testDate, 01.984, 1984, "1984", new BjsAnyObject(bjsObj), bjsObj,
                new Integer[]{1,2,3}, () -> "lambda return value");
        assertEquals("called paramsFunc with params: true, 1984-12-26T00:00:00.000Z, 1.984, 1984, 1984, any object, any object, [1,2,3], lambda return value", ToyClass1.log_static());
    }

    @Test
    public void testParamsFunc_null() {
        ToyClass1.paramsFunc_static(null, null, null, null, null, ToyClass1.bjs.anyNull(), null, null, null);
        assertEquals("called paramsFunc with params: null, null, null, null, null, null, null, null, null", ToyClass1.log_static());
    }

    @Test
    public void testBoolFunc() {
        assertNull(ToyClass1.boolFunc_static());
        assertEquals("called retValueFunc", ToyClass1.log_static());

        ToyClass1.evalAndSetValue_static("null");
        assertNull(ToyClass1.boolFunc_static());

        ToyClass1.evalAndSetValue_static("false");
        assertFalse(ToyClass1.boolFunc_static());

        ToyClass1.evalAndSetValue_static("true");
        assertTrue(ToyClass1.boolFunc_static());

        ToyClass1.evalAndSetValue_static("!!1");
        assertTrue(ToyClass1.boolFunc_static());
    }

    @Test
    public void testDateFunc() {
        assertNull(ToyClass1.dateFunc_static());
        assertEquals("called retValueFunc", ToyClass1.log_static());

        ToyClass1.evalAndSetValue_static("null");
        assertNull(ToyClass1.dateFunc_static());

        ToyClass1.evalAndSetValue_static("new Date(" + testDateTimestamp + ")");
        assertEquals(testDate, ToyClass1.dateFunc_static());
    }

    @Test
    public void testFloatFunc() {
        assertNull(ToyClass1.floatFunc_static());
        assertEquals("called retValueFunc", ToyClass1.log_static());

        ToyClass1.evalAndSetValue_static("null");
        assertNull(ToyClass1.floatFunc_static());

        ToyClass1.evalAndSetValue_static(".1984");
        assertEquals(0.1984, ToyClass1.floatFunc_static(), 0.0);

        ToyClass1.evalAndSetValue_static("123456789.123456789");
        assertEquals(123456789.123456789, ToyClass1.floatFunc_static(), 0.0);
    }

    @Test
    public void testIntFunc() {
        assertNull(ToyClass1.intFunc_static());
        assertEquals("called retValueFunc", ToyClass1.log_static());

        ToyClass1.evalAndSetValue_static("null");
        assertNull(ToyClass1.intFunc_static());

        ToyClass1.evalAndSetValue_static("1984");
        assertEquals(1984, (long) ToyClass1.intFunc_static());

        ToyClass1.evalAndSetValue_static("1234567890");
        assertEquals(1234567890, (long) ToyClass1.intFunc_static());
    }

    @Test
    public void testStringFunc() {
        assertNull(ToyClass1.stringFunc_static());
        assertEquals("called retValueFunc", ToyClass1.log_static());

        ToyClass1.evalAndSetValue_static("null");
        assertNull(ToyClass1.stringFunc_static());

        ToyClass1.evalAndSetValue_static("'hello 84! 👋'");
        assertEquals("hello 84! 👋", ToyClass1.stringFunc_static());
    }

    @Test
    public void testLambdaVoidFunc() {
        Boolean[] check = new Boolean[] {false};

        Lambda.F0<Void> jsLambda = ToyClass1.lambdaVoidFunc_static(() -> {
           assertEquals("calling lambda", ToyClass1.log_static());
           check[0] = true;
           return null;
        });

        assertEquals("called lambdaVoidFunc", ToyClass1.log_static());
        jsLambda.apply();
        assertTrue(check[0]);

        Lambda.F0<Void> nilLambda = ToyClass1.lambdaVoidFunc_static(null);
        assertNull(nilLambda);
    }

    @Test
    public void testLambdaWithParamsFunc() {
        ToyClass1.lambdaWithParamsFunc_static((integer, nullStr, any, obj) -> {
            assertEquals("called lambdaWithParamsFunc", ToyClass1.log_static());
            assertEquals(1984, (int) integer);
            assertNull(nullStr);
            assertEquals("hello 84", any.getObject(ToyClass1.bjsFactory, ToyClass1.class).log());
            assertEquals("hello 84", obj.log());
            return "1984!";
        });
        assertEquals("called lambda with result: 1984!", ToyClass1.log_static());
    }

    @Test
    public void testReturningLambdaWithParamsFunc() {
        Lambda.F4<Integer, BjsAnyObject, ToyClass1, Integer[], String> lambda = ToyClass1.returningLambdaWithParamsFunc_static();
        assertEquals("called returningLambdaWithParamsFunc", ToyClass1.log_static());
        String lambdaRetValue = lambda.apply(1984, new BjsAnyObject(bjsObj), bjsObj, new Integer[]{1,2,3});
        assertEquals("called returned lambda with params: 1984, any object, any object, [1,2,3]", ToyClass1.log_static());
        assertEquals("lambda returning value", lambdaRetValue);
    }

    @Test
    public void testReturningLambdaWithParamsFunc_null() {
        Lambda.F4<Integer, BjsAnyObject, ToyClass1, Integer[], String> lambda = ToyClass1.returningLambdaWithParamsFunc_static();
        assertEquals("called returningLambdaWithParamsFunc", ToyClass1.log_static());
        String lambdaRetValue = lambda.apply(null, ToyClass1.bjs.anyNull(), null, null);
        assertEquals("called returned lambda with params: null, null, null, null", ToyClass1.log_static());
        assertEquals("lambda returning value", lambdaRetValue);
    }
}

File: bionic-js-swift/BjsTests/Tests/ToyClass1Test.swift
import XCTest
import JavaScriptCore

class ToyClass1Test: XCTestCase {
    
    var testDate: Date!
    var testDateTimestamp: Double!
    var bjsObj: ToyClass1!
    
    override func setUp() {
        super.setUp()
        testDateTimestamp = 472867200
        testDate = Date(timeIntervalSince1970: testDateTimestamp)
        bjsObj = ToyClass1()
        bjsObj.log = "any object"
        ToyClass1.evalAndSetValue("undefined")
    }
    
    
    // STATIC PROPERTIES
    
    func testNativeAutoProp() {
        XCTAssertNil(ToyClass1.nativeAutoProp)
        
        ToyClass1.nativeAutoProp = "1984!"
        XCTAssertEqual(ToyClass1.nativeAutoProp, "1984!")
        
        ToyClass1.nativeAutoProp = nil
        XCTAssertNil(ToyClass1.nativeAutoProp)
    }
    
    func testAnyAutoProp() {
        XCTAssertTrue(ToyClass1.anyAutoProp.jsObj.isUndefined)
        
        ToyClass1.anyAutoProp = BjsAnyObject(bjsObj)
        let subclassInstance = ToyClass1.anyAutoProp.getObject(ToyClass2.bjsFactory)!
        let superclassInstance = ToyClass1.anyAutoProp.getObject(ToyClass1.bjsFactory)!
        
        XCTAssertTrue(subclassInstance == superclassInstance)
        XCTAssertTrue(subclassInstance !== superclassInstance)
        
        XCTAssertEqual(subclassInstance.log, "any object")
    }
    
    func testBjsObjAutoProp() {
        XCTAssertNil(ToyClass1.bjsObjAutoProp)
        
        ToyClass1.bjsObjAutoProp = bjsObj
        XCTAssertEqual(ToyClass1.bjsObjAutoProp!.log, "any object")
    }
    
    func testLambdaAutoProp() {
        XCTAssertNil(ToyClass1.lambdaAutoProp)
        
        ToyClass1.lambdaAutoProp = { string in
            if string == nil {
                return "null!"
            }
            return string! + " 1984!"
        }
        XCTAssertEqual(ToyClass1.lambdaAutoProp!(nil), "null!")
        XCTAssertEqual(ToyClass1.lambdaAutoProp!("W"), "W 1984!")
        
        ToyClass1.lambdaAutoProp = nil
        XCTAssertNil(ToyClass1.lambdaAutoProp)
        
        ToyClass1.lambdaAutoProp = { string in
            return nil
        }
        XCTAssertEqual(ToyClass1.lambdaAutoProp!("x"), nil)
    }
    
    func testLambdaAutoProp_putNilToJs() {
        let nullTest = "(this.lambdaAutoProp() === null)"
        
        ToyClass1.lambdaAutoProp = { string in return "hey" }
        ToyClass1.evalAndSetValue(nullTest)
        XCTAssertEqual(ToyClass1.boolFunc(), false)
        
        ToyClass1.lambdaAutoProp = { string in return nil }
        ToyClass1.evalAndSetValue(nullTest)
        XCTAssertEqual(ToyClass1.boolFunc(), true)
    }
    
    func testNativeArrayAutoProp() {
        XCTAssertNil(ToyClass1.nativeArrayAutoProp)
        
        ToyClass1.nativeArrayAutoProp = [[["a", "b"]]]
        XCTAssertEqual(ToyClass1.nativeArrayAutoProp, [[["a", "b"]]])
        
        ToyClass1.nativeArrayAutoProp = [[[], []]]
        XCTAssertEqual(ToyClass1.nativeArrayAutoProp, [[[], []]])
        
        ToyClass1.nativeArrayAutoProp = [[], []]
        XCTAssertEqual(ToyClass1.nativeArrayAutoProp, [[], []])
        
        ToyClass1.nativeArrayAutoProp = []
        XCTAssertEqual(ToyClass1.nativeArrayAutoProp, [])
        
        ToyClass1.nativeArrayAutoProp = [nil, [nil], [[nil]], [["test", nil]]]
        XCTAssertEqual(ToyClass1.nativeArrayAutoProp, [nil, [nil], [[nil]], [["test", nil]]])
    }
    
    func testBjsObjArrayAutoProp() {
        XCTAssertNil(ToyClass1.bjsObjArrayAutoProp)
        
        ToyClass1.bjsObjArrayAutoProp = [[[bjsObj]]]
        XCTAssertEqual(ToyClass1.bjsObjArrayAutoProp, [[[bjsObj]]])
        
        ToyClass1.bjsObjArrayAutoProp = [[[], []]]
        XCTAssertEqual(ToyClass1.bjsObjArrayAutoProp, [[[], []]])
        
        ToyClass1.bjsObjArrayAutoProp = [[], []]
        XCTAssertEqual(ToyClass1.bjsObjArrayAutoProp, [[], []])
        
        ToyClass1.bjsObjArrayAutoProp = []
        XCTAssertEqual(ToyClass1.bjsObjArrayAutoProp, [])
        
        ToyClass1.bjsObjArrayAutoProp = [nil, [nil], [[nil]], [[bjsObj, nil]]]
        XCTAssertEqual(ToyClass1.bjsObjArrayAutoProp, [nil, [nil], [[nil]], [[bjsObj, nil]]])
    }
    
    func testAnyArrayAutoProp() {
        XCTAssertNil(ToyClass1.anyArrayAutoProp)
        
        let anyObj = BjsAnyObject(bjsObj.bjsObj)
        
        ToyClass1.anyArrayAutoProp = [[[ anyObj ]]]
        XCTAssertEqual(ToyClass1.anyArrayAutoProp![0]![0]![0].getObject(ToyClass1.bjsFactory)?.log, "any object")
        
        ToyClass1.anyArrayAutoProp = [[[ ]]]
        XCTAssertTrue(ToyClass1.anyArrayAutoProp![0]![0]!.isEmpty)
    }
    
    func testLambdaArrayAutoProp() {
        XCTAssertNil(ToyClass1.lambdaArrayAutoProp)
        
        let lambda1a: (_ inStr: String?) -> String? = { ($0 ?? "") + "-1a"}
        let lambda1b: (_ inStr: String?) -> String? = { ($0 ?? "") + "-2a"}
        let lambda2a: (_ inStr: String?) -> String? = { ($0 ?? "") + "-1b"}
        
        ToyClass1.lambdaArrayAutoProp = [[[ lambda1a, lambda1b ], [lambda2a, nil] ]]
        
        XCTAssertEqual(ToyClass1.lambdaArrayAutoProp![0]![0]![0]!("test"), "test-1a")
        XCTAssertEqual(ToyClass1.lambdaArrayAutoProp![0]![0]![1]!("test"), "test-2a")
        XCTAssertEqual(ToyClass1.lambdaArrayAutoProp![0]![1]![0]!("test"), "test-1b")
        XCTAssertNil(ToyClass1.lambdaArrayAutoProp![0]![1]![1])
        
        ToyClass1.lambdaArrayAutoProp = [[[ ]]]
        XCTAssertTrue(ToyClass1.lambdaArrayAutoProp![0]![0]!.isEmpty)
    }
    
    func testProp() {
        XCTAssertEqual(ToyClass1.prop, "1984!")
        
        ToyClass1.prop = "test value"
        XCTAssertEqual(ToyClass1.prop, "test value")
        
        ToyClass1.prop = nil
        XCTAssertEqual(ToyClass1.prop, "1984!")
    }
    
    
    // STATIC METHODS
    
    func testVoidFunc() {
        ToyClass1.voidFunc()
        XCTAssertEqual(ToyClass1.log, "called voidFunc")
    }
    
    func testParamsFunc() {
        let testExpectation = expectation(description: "called")
        ToyClass1.paramsFunc(true, testDate, 01.984, 01984, "1984", BjsAnyObject(bjsObj.bjsObj), bjsObj,
                             [1,2,3]) { () -> String? in
            testExpectation.fulfill()
            return "lambda return value"
        }
        XCTAssertEqual(ToyClass1.log, "called paramsFunc with params: true, 1984-12-26T00:00:00.000Z, 1.984, 1984, 1984, any object, any object, [1,2,3], lambda return value")
        wait(for: [testExpectation], timeout: 10)
    }
    
    func testParamsFunc_nil() {
        ToyClass1.paramsFunc(nil, nil, nil, nil, nil, ToyClass1.bjs.anyNull, nil, nil, nil)
        XCTAssertEqual(ToyClass1.log, "called paramsFunc with params: null, null, null, null, null, null, null, null, null")
    }
    
    func testBoolFunc() {
        XCTAssertNil(ToyClass1.boolFunc())
        XCTAssertEqual(ToyClass1.log, "called retValueFunc")
        
        ToyClass1.evalAndSetValue("null")
        XCTAssertNil(ToyClass1.boolFunc())
        
        ToyClass1.evalAndSetValue("false")
        XCTAssertEqual(ToyClass1.boolFunc(), false)
        
        ToyClass1.evalAndSetValue("true")
        XCTAssertEqual(ToyClass1.boolFunc(), true)
        
        ToyClass1.evalAndSetValue("1")
        XCTAssertEqual(ToyClass1.boolFunc(), true)
    }
    
    func testDateFunc() {
        XCTAssertNil(ToyClass1.dateFunc())
        XCTAssertEqual(ToyClass1.log, "called retValueFunc")
        
        ToyClass1.evalAndSetValue("null")
        XCTAssertNil(ToyClass1.dateFunc())
        
        ToyClass1.evalAndSetValue("new Date(\(testDateTimestamp!) * 1000)")
        XCTAssertEqual(ToyClass1.dateFunc(), testDate)
    }
    
    func testFloatFunc() {
        XCTAssertNil(ToyClass1.floatFunc())
        XCTAssertEqual(ToyClass1.log, "called retValueFunc")
        
        ToyClass1.evalAndSetValue("null")
        XCTAssertNil(ToyClass1.floatFunc())
        
        ToyClass1.evalAndSetValue(".1984")
        XCTAssertEqual(ToyClass1.floatFunc(), 0.1984)
        
        ToyClass1.evalAndSetValue("123456789.123456789")
        XCTAssertEqual(ToyClass1.floatFunc(), 123456789.123456789)
    }
    
    func testIntFunc() {
        XCTAssertNil(ToyClass1.intFunc())
        XCTAssertEqual(ToyClass1.log, "called retValueFunc")
        
        ToyClass1.evalAndSetValue("null")
        XCTAssertNil(ToyClass1.intFunc())
        
        ToyClass1.evalAndSetValue("1984")
        XCTAssertEqual(ToyClass1.intFunc(), 1984)
        
        ToyClass1.evalAndSetValue("1234567890")
        XCTAssertEqual(ToyClass1.intFunc(), 1234567890)
    }
    
    func testStringFunc() {
        XCTAssertNil(ToyClass1.stringFunc())
        XCTAssertEqual(ToyClass1.log, "called retValueFunc")
        
        ToyClass1.evalAndSetValue("null")
        XCTAssertNil(ToyClass1.stringFunc())
        
        ToyClass1.evalAndSetValue("'hello 84! 👋'")
        XCTAssertEqual(ToyClass1.stringFunc(), "hello 84! 👋")
    }
    
    func testLambdaVoidFunc() {
        let testExpectation = expectation(description: "called")
        
        let jsLambda = ToyClass1.lambdaVoidFunc {
            XCTAssertEqual(ToyClass1.log, "calling lambda")
            testExpectation.fulfill()
        }
        XCTAssertEqual(ToyClass1.log, "called lambdaVoidFunc")
        jsLambda!()
        wait(for: [testExpectation], timeout: 10)
        
        let nilLambda = ToyClass1.lambdaVoidFunc(nil)
        XCTAssertNil(nilLambda)
    }
    
    func testLambdaVoidFunc_nil() {
        let nilLambda = ToyClass1.lambdaVoidFunc(nil)
        XCTAssertNil(nilLambda)
    }
    
    func testLambdaWithParamsFunc() {
        let testExpectation = expectation(description: "called")
        ToyClass1.lambdaWithParamsFunc { (int, nullStr, any, obj) in
            XCTAssertEqual(ToyClass1.log, "called lambdaWithParamsFunc")
            XCTAssertEqual(int, 1984)
            XCTAssertNil(nullStr)
            XCTAssertEqual(any.getObject(ToyClass1.bjsFactory)?.log, "hello 84")
            XCTAssertEqual(obj?.log, "hello 84")
            testExpectation.fulfill()
            return "1984!"
        }
        XCTAssertEqual(ToyClass1.log, "called lambda with result: 1984!")
        wait(for: [testExpectation], timeout: 10)
    }
    
    func testReturningLambdaWithParamsFunc() {
        let lambda = ToyClass1.returningLambdaWithParamsFunc()
        XCTAssertEqual(ToyClass1.log, "called returningLambdaWithParamsFunc")
        let lambdaRetValue = lambda!(1984, BjsAnyObject(bjsObj.bjsObj), bjsObj, [1,2,3])
        XCTAssertEqual(ToyClass1.log, "called returned lambda with params: 1984, any object, any object, [1,2,3]")
        XCTAssertEqual(lambdaRetValue, "lambda returning value")
    }
    
    func testReturningLambdaWithParamsFunc_nil() {
        let lambda = ToyClass1.returningLambdaWithParamsFunc()
        XCTAssertEqual(ToyClass1.log, "called returningLambdaWithParamsFunc")
        let lambdaRetValue = lambda!(nil, ToyClass1.bjs.anyNull, nil, nil)
        XCTAssertEqual(ToyClass1.log, "called returned lambda with params: null, null, null, null")
        XCTAssertEqual(lambdaRetValue, "lambda returning value")
    }
}
